### DISCUSSION TITLE
Migrate spawn from node to bun

### DISCUSSION THREAD

**Discussion by equiman** [oven-sh/bun#5034] [Q&A]
I'm been trying to migrate a cli utility from `node` to `bun`. It's a tool that receives a command with arguments and run it using `spawn`. 

Original method with node:

```js
import { exit } from 'node:process'
import { spawn } from 'node:child_process'
import chalk from 'chalk'
import { stripIndents } from 'common-tags'

export const spreadCommand = async (cmd: string, args: string[]) => {
  const child = spawn(
    cmd,
    args,
    {
      stdio: 'inherit',
      shell: true
    }
  )

  child.on('error', (error) => {
    console.error(stripIndents`
        ${chalk.red.bold('Error')}:
        ${error}
      `)
    exit(1)
  })

  child.on('exit', (code) => {
    return code
  })
}
```

Method migrated to Bun:

```js
import chalk from 'chalk'
import { stripIndents } from 'common-tags'

export const spreadCommand = async (cmd: string, args: string[]) => {

  const proc = Bun.spawnSync(
    [cmd, ...args],
    {
      stdio: ['inherit', 'inherit', 'inherit'],
      shell: true,
      onExit(proc, exitCode, signalCode, error) {
        if (error) {
          console.error(stripIndents`
            ${chalk.red.bold('Error')}:
            ${proc.stderr}
          `)

          proc.kill(1)
        }
      },
    }
  )

  return proc.exitCode
}
```

---

I've tested it. When the command exist (like 'bun install') all works fine, but when the commant doesn't exist it will show a bun error instead of showing the command error.

```js
const value = await spreadCommand('bunjs', ['install'])
console.log({value})
```

The code never go inside the `onError`, intead the function fails:
![image](https://github.com/oven-sh/bun/assets/933393/d73fa527-9464-434b-a52b-390130f4a4e1)

Same test with previous method written in node show the expected error in the terminal:
![image](https://github.com/oven-sh/bun/assets/933393/fea3067f-edcf-42ad-b889-0bed72e851e6)

# Docs

- https://bun.sh/docs/api/spawn#blocking-api-bun-spawnsync
- https://bun.sh/guides/process/spawn
- https://bun.sh/guides/process/spawn-stdout
- https://bun.sh/guides/process/spawn-stderr

**Comment by equiman** [comment]
I've also tried this other way, but I got the same result.

```js
import chalk from 'chalk'
import { stripIndents } from 'common-tags'

export const spreadCommand = async (cmd: string, args: string[]) => {
  const proc = Bun.spawn([cmd, ...args],
    {
      stdin: 'inherit',
      stdout: 'inherit',
      stderr: 'pipe',
      shell: true
    });

  const errors: string = await Bun.readableStreamToText(proc.stderr);
  if (errors) {
    console.error(stripIndents`
      ${chalk.red.bold('Error')}:
      ${errors}
    `)

    proc.kill(1)
  }

  return proc.exitCode
}
```

**Reply by calumk** [reply]
Did you resolve this? 

Spawn seems a little bit unfinished 

It seems to be very hard to catch stdout and stderr properly


I'm trying to run a command and catch all outputs and publish them via web socket and it seems much easier using the node .on() approach
